File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] max = new int[nums2.length];
        Stack<Integer> stack = new Stack<>(); //keeps track of next max
        int count = nums2.length-1;
        while(count>=0){
            if(stack.empty()){
                stack.push(nums2[count]);
                max[count] = -1;
            }
            else{
                if(stack.peek()>nums2[count]){
                    max[count]= stack.peek();
                    stack.push(nums2[count]);
                }
                else{
                    while(!stack.empty() && stack.peek()<nums2[count]){
                        stack.pop();
                    }
                    if(stack.empty()){
                        max[count]=-1;
                    }
                    else{
                        max[count]= stack.peek();
                    }
                    stack.push(nums2[count]);
                }
            }
            count--;
        }
        int[] ans = new int[nums1.length];
        for(int i =0; i<nums2.length; i++){
            for(int j=0; j<nums1.length; j++){
                if(nums1[j]==nums2[i]){
                    ans[j]=max[i];
                }
            }
        }
        return ans;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int len = nums2.size();
        map<int,int> nge;
        vector<int> ans(nums1.size(), 0);
        stack<int> st;
        nge[nums2[len-1]] = -1;
        st.push(nums2[len-1]);
        for(int i = len-2; i>=0; i--){
            if(st.top() > nums2[i]){
                nge[nums2[i]] = st.top();
                st.push(nums2[i]);
            }
            else{
                while(!st.empty() && st.top()< nums2[i]){
                    st.pop();
                }
                if(st.empty()) nge[nums2[i]] = -1;
                else nge[nums2[i]] = st.top();

                st.push(nums2[i]);
            }
        }
        for(int i=0; i<nums1.size(); i++){
            ans[i] = nge[nums1[i]];
        }
        return ans;
    }
};